### PR TITLE
[dynamo] mirror training flag in OptimizedModule

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131546

Fixes https://github.com/pytorch/pytorch/issues/122414.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by yanboliang — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -197,6 +197,11 @@ def __setstate__(self, state):
         self.__dict__ = state
         self._initialize()
 
+    def __getattribute__(self, name):
+        if name == "training":
+            return getattr(self._orig_mod, name)
+        return super().__getattribute__(name)
+
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:204]: Perhaps:
```suggestion
    @property
    def training(self):
        return getattr(self._orig_mod, name)
```

The prior method will add overhead on every single attribute access.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/eval_frame.py]: Needs a little more work to get the setter working.

**Pull request closed without merging**